File: frontend/src/types/notification.ts
/**
 * 通知型別
 */

export type NotificationType =
    | 'low_stock'
    | 'expiry_warning'
    | 'document_approval'
    | 'protocol_status'
    | 'protocol_submitted'
    | 'vet_recommendation'
    | 'system_alert'
    | 'monthly_report'
    | 'leave_approval'
    | 'review_assignment'
    | 'review_comment'

export const notificationTypeNames: Record<NotificationType, string> = {
    low_stock: '低庫存預警',
    expiry_warning: '效期預警',
    document_approval: '單據審核',
    protocol_status: '計畫狀態',
    protocol_submitted: '計畫送審',
    vet_recommendation: '獸醫師建議',
    system_alert: '系統通知',
    monthly_report: '月報',
    leave_approval: '請假審核',
    review_assignment: '審查指派',
    review_comment: '審查意見',
}

export interface NotificationItem {
    id: string
    type: NotificationType
    title: string
    content?: string
    is_read: boolean
    read_at?: string
    related_entity_type?: string
    related_entity_id?: string
    created_at: string
    /** 0=一般；1=緊急置頂（待辦，完成前排在清單最上方） */
    priority?: number
}

export interface NotificationListResponse {
    data: NotificationItem[]
    total: number
    page: number
    per_page: number
}

export interface UnreadNotificationCount {
    count: number
}

export interface MarkNotificationsReadRequest {
    notification_ids: string[]
}

// 通知設定
export interface NotificationSettings {
    user_id: string
    email_low_stock: boolean
    email_expiry_warning: boolean
    email_document_approval: boolean
    email_protocol_status: boolean
    email_monthly_report: boolean
    expiry_warning_days: number
    low_stock_notify_immediately: boolean
    updated_at: string
}

export interface UpdateNotificationSettingsRequest {
    email_low_stock?: boolean
    email_expiry_warning?: boolean
    email_document_approval?: boolean
    email_protocol_status?: boolean
    email_monthly_report?: boolean
    expiry_warning_days?: number
    low_stock_notify_immediately?: boolean
}

// ============================================
// 通知路由規則
// ============================================

export type NotificationFrequency = 'immediate' | 'daily' | 'weekly' | 'monthly'

export interface NotificationRouting {
    id: string
    event_type: string
    role_code: string
    channel: string       // 'in_app' | 'email' | 'both'
    is_active: boolean
    description?: string
    /** 批次通知頻率 */
    frequency: NotificationFrequency
    /** 批次通知執行小時（0-23） */
    hour_of_day: number
    /** weekly 時有效：0=週日, 1=週一 ... 6=週六 */
    day_of_week: number | null
    created_at: string
    updated_at: string
}

export interface CreateNotificationRoutingRequest {
    event_type: string
    role_code: string
    channel?: string
    description?: string
    frequency?: NotificationFrequency
    hour_of_day?: number
    day_of_week?: number | null
}

export interface UpdateNotificationRoutingRequest {
    channel?: string
    is_active?: boolean
    description?: string
    frequency?: NotificationFrequency
    hour_of_day?: number
    day_of_week?: number | null
}

// ============================================
// 效期通知範圍設定（系統層級）
// ============================================

export interface ExpiryNotificationConfig {
    id: string
    /** 提前幾天開始預警（預設 60） */
    warn_days: number
    /** 過期超過幾天後停止通知（預設 90） */
    cutoff_days: number
    /** 過期超過此天數後轉月度彙整通知；null=停用 */
    monthly_threshold_days: number | null
    updated_at: string
    updated_by: string | null
}

export interface UpdateExpiryNotificationConfigRequest {
    warn_days?: number
    cutoff_days?: number
    /** null = 停用月度模式 */
    monthly_threshold_days?: number | null
}

/** 事件類型資訊 */
export interface EventTypeInfo {
    code: string
    name: string
}

/** 事件類型分類（含主要分組 AUP | Animal | ERP | HR） */
export interface EventTypeCategory {
    group: string
    category: string
    event_types: EventTypeInfo[]
}

/** 角色資訊 */
export interface RoleInfo {
    code: string
    name: string
}

/** 事件類型中文名稱對照 */
export const eventTypeNames: Record<string, string> = {
    protocol_submitted: '計畫提交',
    protocol_vet_review: '獸醫審查',
    protocol_under_review: '委員審查',
    protocol_resubmitted: '重新提交',
    protocol_approved: '計畫核准',
    protocol_rejected: '計畫駁回',
    review_comment_created: '新審查意見',
    all_reviews_completed: '所有審查意見送出',
    all_comments_resolved: '所有意見已解決',
    leave_submitted: '請假申請',
    overtime_submitted: '加班申請',
    leave_approved: '請假核准',
    overtime_approved: '加班核准',
    document_submitted: '採購單提交',
    low_stock_alert: '低庫存預警',
    expiry_alert: '效期預警',
    emergency_medication: '緊急給藥',
    animal_abnormal_record: '動物異常紀錄',
    vet_recommendation_created: '獸醫師建議',
    animal_sudden_death: '動物猝死',
    euthanasia_order_created: '安樂死申請',
    amendment_submitted: '修正案提交',
    amendment_decision_recorded: '修正案審查決定',
    amendment_approved: '修正案核准',
    amendment_rejected: '修正案駁回',
    leave_cancelled: '請假取消',
    po_pending_receipt: '採購單未入庫提醒',
}

/** 通道中文名稱對照 */
export const channelNames: Record<string, string> = {
    in_app: '站內通知',
    email: 'Email',
    both: '兩者',
}

/** 頻率中文名稱對照 */
export const frequencyNames: Record<NotificationFrequency, string> = {
    immediate: '即時',
    daily: '每日',
    weekly: '每週',
    monthly: '每月',
}

/** 可設定批次頻率的事件類型（非 event-driven） */
export const BATCH_EVENT_TYPES = new Set([
    'expiry_alert',
    'low_stock_alert',
    'po_pending_receipt',
    'equipment_overdue',
])

File: backend/src/models/notification.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use utoipa::ToSchema;
use uuid::Uuid;
use validator::Validate;

/// 通知類型
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, sqlx::Type)]
#[sqlx(type_name = "notification_type", rename_all = "snake_case")]
pub enum NotificationType {
    LowStock,
    ExpiryWarning,
    DocumentApproval,
    ProtocolStatus,
    ProtocolSubmitted,
    ReviewAssignment,
    ReviewComment,
    LeaveApproval,
    OvertimeApproval,
    VetRecommendation,
    SystemAlert,
    MonthlyReport,
}

impl NotificationType {
    pub fn as_str(&self) -> &'static str {
        match self {
            NotificationType::LowStock => "low_stock",
            NotificationType::ExpiryWarning => "expiry_warning",
            NotificationType::DocumentApproval => "document_approval",
            NotificationType::ProtocolStatus => "protocol_status",
            NotificationType::ProtocolSubmitted => "protocol_submitted",
            NotificationType::ReviewAssignment => "review_assignment",
            NotificationType::ReviewComment => "review_comment",
            NotificationType::LeaveApproval => "leave_approval",
            NotificationType::OvertimeApproval => "overtime_approval",
            NotificationType::VetRecommendation => "vet_recommendation",
            NotificationType::SystemAlert => "system_alert",
            NotificationType::MonthlyReport => "monthly_report",
        }
    }
}

/// 通知
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Notification {
    pub id: Uuid,
    pub user_id: Uuid,
    pub r#type: String,
    pub title: String,
    pub content: Option<String>,
    pub is_read: bool,
    pub read_at: Option<DateTime<Utc>>,
    pub related_entity_type: Option<String>,
    pub related_entity_id: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    /// 0=一般；1=緊急置頂（待辦，完成前排在清單最上方）
    pub priority: i16,
}

/// 通知優先級：一般（隨時間新到舊排序）。
pub const PRIORITY_NORMAL: i16 = 0;
/// 通知優先級：緊急置頂（待辦，完成前恆排在清單最上方；完成後由 hook 降級回 [`PRIORITY_NORMAL`]）。
pub const PRIORITY_PINNED: i16 = 1;

/// 通知列表項目（含額外資訊）
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct NotificationItem {
    pub id: Uuid,
    pub r#type: String,
    pub title: String,
    pub content: Option<String>,
    pub is_read: bool,
    pub read_at: Option<DateTime<Utc>>,
    pub related_entity_type: Option<String>,
    pub related_entity_id: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    /// 0=一般；1=緊急置頂（待辦，完成前排在清單最上方）
    pub priority: i16,
}

/// 通知設定
#[derive(Debug, Clone, Serialize, Deserialize, FromRow, ToSchema)]
pub struct NotificationSettings {
    pub user_id: Uuid,
    pub email_low_stock: bool,
    pub email_expiry_warning: bool,
    pub email_document_approval: bool,
    pub email_protocol_status: bool,
    pub email_monthly_report: bool,
    pub expiry_warning_days: i32,
    pub low_stock_notify_immediately: bool,
    pub updated_at: DateTime<Utc>,
}

/// 更新通知設定請求
#[derive(Debug, Deserialize, Validate)]
pub struct UpdateNotificationSettingsRequest {
    pub email_low_stock: Option<bool>,
    pub email_expiry_warning: Option<bool>,
    pub email_document_approval: Option<bool>,
    pub email_protocol_status: Option<bool>,
    pub email_monthly_report: Option<bool>,
    #[validate(range(min = 1, max = 90, message = "Expiry warning days must be 1-90"))]
    pub expiry_warning_days: Option<i32>,
    pub low_stock_notify_immediately: Option<bool>,
}

/// 建立通知請求（內部使用）
#[derive(Debug, Deserialize)]
pub struct CreateNotificationRequest {
    pub user_id: Uuid,
    pub notification_type: NotificationType,
    pub title: String,
    pub content: Option<String>,
    pub related_entity_type: Option<String>,
    pub related_entity_id: Option<Uuid>,
}

/// 通知查詢參數
#[derive(Debug, Deserialize)]
pub struct NotificationQuery {
    pub is_read: Option<bool>,
    pub notification_type: Option<String>,
}

/// 未讀通知數量
#[derive(Debug, Serialize)]
pub struct UnreadNotificationCount {
    pub count: i64,
}

/// 標記已讀請求
#[derive(Debug, Deserialize)]
pub struct MarkNotificationsReadRequest {
    pub notification_ids: Vec<Uuid>,
}

// LowStockAlert is defined in models/stock.rs to avoid duplication

/// 效期預警項目
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct ExpiryAlert {
    pub product_id: Uuid,
    pub sku: String,
    pub product_name: String,
    pub spec: Option<String>,
    pub category_code: Option<String>,
    pub warehouse_id: Uuid,
    pub warehouse_code: String,
    pub warehouse_name: String,
    pub batch_no: Option<String>,
    pub expiry_date: chrono::NaiveDate,
    pub on_hand_qty: rust_decimal::Decimal,
    pub base_uom: String,
    pub days_until_expiry: i32,
    pub expiry_status: String,
    /// 同品項同倉庫所有批號的合計庫存量
    pub total_qty: rust_decimal::Decimal,
}

// ============================================
// 定期報表相關
// ============================================

/// 排程類型
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, sqlx::Type)]
#[sqlx(type_name = "schedule_type", rename_all = "lowercase")]
pub enum ScheduleType {
    Daily,
    Weekly,
    Monthly,
}

/// 報表類型
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, sqlx::Type)]
#[sqlx(type_name = "report_type", rename_all = "snake_case")]
pub enum ReportType {
    StockOnHand,
    StockLedger,
    PurchaseSummary,
    CostSummary,
    ExpiryReport,
    LowStockReport,
}

/// 定期報表設定
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct ScheduledReport {
    pub id: Uuid,
    pub report_type: String,
    pub schedule_type: String,
    pub day_of_week: Option<i32>,
    pub day_of_month: Option<i32>,
    pub hour_of_day: i32,
    pub parameters: Option<serde_json::Value>,
    pub recipients: Vec<Uuid>,
    pub is_active: bool,
    pub last_run_at: Option<DateTime<Utc>>,
    pub next_run_at: Option<DateTime<Utc>>,
    pub created_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// 報表歷史記錄
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct ReportHistory {
    pub id: Uuid,
    pub scheduled_report_id: Option<Uuid>,
    pub report_type: String,
    pub file_name: String,
    pub file_path: String,
    pub file_size: Option<i32>,
    pub parameters: Option<serde_json::Value>,
    pub generated_at: DateTime<Utc>,
    pub generated_by: Option<Uuid>,
}

/// 建立定期報表請求
#[derive(Debug, Deserialize, Validate)]
pub struct CreateScheduledReportRequest {
    pub report_type: String,
    pub schedule_type: String,
    #[validate(range(min = 0, max = 6, message = "Day of week must be 0-6"))]
    pub day_of_week: Option<i32>,
    #[validate(range(min = 1, max = 31, message = "Day of month must be 1-31"))]
    pub day_of_month: Option<i32>,
    #[validate(range(min = 0, max = 23, message = "Hour must be 0-23"))]
    pub hour_of_day: i32,
    pub parameters: Option<serde_json::Value>,
    #[validate(length(min = 1, message = "At least one recipient required"))]
    pub recipients: Vec<Uuid>,
}

/// 更新定期報表請求
#[derive(Debug, Deserialize, Validate)]
pub struct UpdateScheduledReportRequest {
    pub day_of_week: Option<i32>,
    pub day_of_month: Option<i32>,
    pub hour_of_day: Option<i32>,
    pub parameters: Option<serde_json::Value>,
    pub recipients: Option<Vec<Uuid>>,
    pub is_active: Option<bool>,
}

// ============================================
// 通知路由規則
// ============================================

/// 通知路由規則（事件→收件人來源→通道的對應）
///
/// 收件人來源由 `target_kind` 決定：
/// - `role`：`target_value` 為 `roles.code`，收件人＝持有該角色者（`role_code` 同步保留供舊路徑）。
/// - `resolver`：`target_value` 為關係型解析器 key（見 `services/notification/resolvers.rs`），
///   收件人由觸發事件的實體關係算出（例：這張假單的申請人、這個計畫的 PI）。`role_code` 為 NULL。
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct NotificationRouting {
    pub id: Uuid,
    pub event_type: String,
    /// 角色型列為角色代碼；resolver 型列為 NULL（過渡期保留供 get_recipients_by_event 等舊路徑）
    pub role_code: Option<String>,
    pub channel: String,
    pub is_active: bool,
    pub description: Option<String>,
    /// 批次通知頻率：immediate | daily | weekly | monthly
    pub frequency: String,
    /// 批次通知執行小時（0-23）
    pub hour_of_day: i16,
    /// weekly 時有效：0=週日, 1=週一 ... 6=週六
    pub day_of_week: Option<i16>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    /// 收件人來源種類：`role` | `resolver`
    pub target_kind: String,
    /// `role` 時為 roles.code；`resolver` 時為 resolver key
    pub target_value: String,
}

/// 建立通知路由規則請求
#[derive(Debug, Deserialize, Validate)]
pub struct CreateNotificationRoutingRequest {
    #[validate(length(min = 1, max = 80, message = "event_type 不可為空且最多 80 字元"))]
    pub event_type: String,
    #[validate(length(min = 1, max = 50, message = "role_code 不可為空且最多 50 字元"))]
    pub role_code: String,
    pub channel: Option<String>,
    pub description: Option<String>,
    pub frequency: Option<String>,
    #[validate(range(min = 0, max = 23, message = "hour_of_day 必須介於 0-23"))]
    pub hour_of_day: Option<i16>,
    #[validate(range(min = 0, max = 6, message = "day_of_week 必須介於 0-6"))]
    pub day_of_week: Option<i16>,
}

/// 更新通知路由規則請求
#[derive(Debug, Deserialize)]
pub struct UpdateNotificationRoutingRequest {
    pub channel: Option<String>,
    pub is_active: Option<bool>,
    pub description: Option<String>,
    pub frequency: Option<String>,
    pub hour_of_day: Option<i16>,
    pub day_of_week: Option<i16>,
}

// ============================================
// 效期通知範圍設定
// ============================================

/// 系統層級效期通知範圍設定（全系統單一列）
#[derive(Debug, Clone, Serialize, Deserialize, FromRow, ToSchema)]
pub struct ExpiryNotificationConfig {
    pub id: Uuid,
    /// 提前幾天開始預警（預設 60）
    pub warn_days: i16,
    /// 過期超過幾天後停止通知（預設 90）
    pub cutoff_days: i16,
    /// 過期超過此天數後轉月度彙整通知；None=停用
    pub monthly_threshold_days: Option<i16>,
    pub updated_at: DateTime<Utc>,
    pub updated_by: Option<Uuid>,
}

/// 更新效期通知範圍設定請求
#[derive(Debug, Deserialize, Validate)]
pub struct UpdateExpiryNotificationConfigRequest {
    #[validate(range(min = 1, max = 365, message = "warn_days 必須介於 1-365"))]
    pub warn_days: Option<i16>,
    #[validate(range(min = 1, max = 730, message = "cutoff_days 必須介於 1-730"))]
    pub cutoff_days: Option<i16>,
    /// None = 停用月度模式
    pub monthly_threshold_days: Option<i16>,
}

/// 事件類型分類（含主要分組、子分類名稱與事件清單）
/// group: AUP | Animal | ERP | HR，供前端分頁/分區顯示
#[derive(Debug, Clone, Serialize)]
pub struct EventTypeCategory {
    /// 主要分組：AUP | Animal | ERP | HR
    pub group: String,
    /// 子分類顯示名稱
    pub category: String,
    pub event_types: Vec<EventTypeInfo>,
}

/// 事件類型資訊
#[derive(Debug, Clone, Serialize)]
pub struct EventTypeInfo {
    pub code: String,
    pub name: String,
}

/// 角色資訊（供通知路由下拉選單使用）
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct RoleInfo {
    pub code: String,
    pub name: String,
}

/// 某事件的「具體收件人」預覽（供 admin 路由頁巢狀收合顯示「實際是誰」）。
#[derive(Debug, Clone, Serialize)]
pub struct RoutingRecipientsPreview {
    pub event_type: String,
    pub rules: Vec<RoutingRuleRecipients>,
}

/// 單一路由規則解析後的收件人資訊。
#[derive(Debug, Clone, Serialize)]
pub struct RoutingRuleRecipients {
    pub rule_id: Uuid,
    pub target_kind: String,
    pub target_value: String,
    pub channel: String,
    pub is_active: bool,
    /// 人類可讀標籤（角色中文名 or resolver 標籤）。
    pub label: String,
    /// resolver 型：對象由事件動態決定的說明；角色型為 None。
    pub description: Option<String>,
    /// 是否「固定收件人、不可調整」（resolver 型為 true）。
    pub fixed: bool,
    /// 角色型：實際持有該角色的使用者；resolver 型為空（依事件動態決定，無法靜態列舉）。
    pub members: Vec<RecipientMember>,
}

/// 收件人成員（角色型規則展開用）。
#[derive(Debug, Clone, Serialize, FromRow)]
pub struct RecipientMember {
    pub id: Uuid,
    pub display_name: String,
    pub email: String,
}

/// 「固定通知」目錄項：收件人由系統流程決定、不經 notification_routing、不可調整。
/// 供 admin 路由頁顯示為唯讀 notes，讓管理者看見「有什麼狀況會通知誰」。
#[derive(Debug, Clone, Serialize)]
pub struct FixedNotification {
    /// 中文事件名稱。
    pub event_label: String,
    /// 觸發情境。
    pub trigger: String,
    /// 收件人（描述，由流程動態決定）。
    pub recipient_label: String,
    /// 通知管道：'in_app' | 'email' | 'both'。
    pub channel: String,
    /// 分組（對齊路由頁分組，如 Animal）。
    pub group: String,
    /// 為何固定 / 補充說明。
    pub note: String,
}
